// tool/lib/remote_node_helper.js
'use strict';

function convertRemoteNodeArray(nodes, highestId) {
    var nodeMap = {
        data: {},
        gui:  {},

        highestId: typeof highestId === 'number' ? highestId : 0
    };

    nodes.forEach(function(node) {
        nodeMap.data[node.id] = {
            id:             node.id,
            syncId:         node.id,
            name:           node.name,
            description:    node.description,
            type:           node.type,
            prototypeId:    node.prototype_id,
            role:           node.role,
            simulateChange: 0,
            baseline:       node.baseline || 0,

            objectId:       'nodeData'
        };

        nodeMap.gui[node.id]  = {
            id:         node.id,
            syncId:     node.id,
            radius:     node.radius,
            x:          node.x,
            y:          node.y,
            avatar:     node.avatar,
            color:      node.color,
            links:      [],

            objectId:   'nodeGui'
        };

        if(nodeMap.highestId < node.id) {
            nodeMap.highestId = node.id;
        }
    });

    return nodeMap;
}

module.exports = {
    convertRemoteNodeArray: convertRemoteNodeArray
};
