// model/model.js
'use strict';

module.exports = {
    listeners: {
        delete:         require('./listeners/delete.js'),
        deleteModel:    require('./listeners/delete_model.js'),
        deleteSelected: require('./listeners/delete_selected.js'),
        loadModel:      require('./listeners/load_model.js'),
        mouseDown:      require('./listeners/mouse_down.js'),
        mouseMove:      require('./listeners/mouse_move.js'),
        mouseUp:        require('./listeners/mouse_up.js'),
        newModel:       require('./listeners/new_model.js'),
        nodeModified:   require('./listeners/node_modified.js'),
        notification:   require('./listeners/notification.js'),
        popup:          require('./listeners/popup.js'),
        resetUI:        require('./listeners/reset_ui.js'),
        saveModel:      require('./listeners/save_model.js'),
        selected:       require('./listeners/selected.js'),
        settings:       require('./listeners/settings.js'),
        storeModel:     require('./listeners/store_model.js')
    }
};
