// new_ui/lib/button.js
'use strict';

var Element = require('./element.js');

function Button() {
    Element.call(this, 'div');

    this.root.style.cursor = 'pointer';
    this.label = new Element('div');

    this.appendChild(this.label);

    this.clicks = [];
}

Button.prototype = {
    setLabel: function(label) {
        this.label.setLabel(label);
    },

    simulateClick: function() {
        this.clicks.forEach(function(click) {
            click();
        });
    },

    click: function(callback) {
        if(!callback || typeof callback !== 'function') {
            return;
        }
        
        this.clicks.push(callback);
        var that = this;
        this.root.addEventListener('click', function(e) {
            that.emit('clicked');
            callback(e);
        });
    },

    removeClick: function(callback) {
        var index = this.clicks.indexOf(callback);
        this.clicks.splice(index, 1);

        this.root.removeEventListener('click', callback);
    },

    removeEvents: function() {
        this.clicks.forEach(function(callback) {
            this.root.removeEventListener('click', callback);
        }, this);

        this.clicks = [];
    },

    replaceClick: function(callback, newCallback) {
        var index = this.clicks.indexOf(callback);
        if(index !== -1) {
            this.root.removeEventListener('click', this.clicks[index]);
            this.clicks[index] = newCallback;
        } else {
            this.clicks.push(newCallback);
        }

        this.root.addEventListener('click', newCallback);
    },

    __proto__: Element.prototype
};

module.exports = Button;
