// new_ui/lib/checkbox.js
'use strict';

var Input  = require('./input.js'),
    Colors = require('./colors.js'),
    Button = require('./button.js');

function Checkbox() {
    Button.call(this);

    this.checked = false;

    this.uncheck;
    this.check;

    var that = this;
    this.click(function() {
        if(that.checked) {
            that.uncheck();
        } else {
            that.check();
        }
    });
}

Checkbox.prototype = {
    setBackground: function(background) {
        this.nonHighlightBackground = background;
        Input.prototype.setBackground.call(this, background);
    },

    onUncheck: function(callback) {
        var that = this;
        that.uncheck = function() {
            Input.prototype.setBackground.call(that, that.nonHighlightBackground);
            callback();
            that.checked = false;
            that.emit('unchecked');
        };
    },

    onCheck: function(callback) {
        var that = this;
        that.check = function() {
            Input.prototype.setBackground.call(that, Colors.buttonCheckedBackground);
            callback();
            that.checked = true;
            that.emit('checked');
        };
    },

    removeEvents: function() {
        Button.prototype.removeEvents.call(this);
    },
    __proto__: Button.prototype
};

module.exports = Checkbox;
