// new_ui/lib/checkbox.js
'use strict';

var Input   = require('./input.js'),
    Colors  = require('./colors.js'),
    Element = require('./element.js'),
    Button  = require('./button.js');

function Checkbox() {
    Button.call(this);

    this.circleContainer = new Element('div');
    this.circleDiv       = new Element('div');

    this.root.style.position                  = 'relative';
    this.circleContainer.root.style.position  = 'absolute';
    this.circleContainer.root.style.top       = '50%';
    this.circleContainer.root.style.transform = 'translateY(-50%)';
    this.circleContainer.root.style.left      = '8px';

    this.circleDiv.setWidth('16px');
    this.circleDiv.setHeight('16px');
    this.circleDiv.root.style.border = '1px solid ' + Colors.buttonCheckboxCircleBorder;
    this.circleDiv.setBackground(Colors.buttonUncheckedCircle);
    this.circleDiv.root.style['border-radius'] = '50%';

    this.circleContainer.appendChild(this.circleDiv);

    this.insertChildAt(this.circleContainer, 0);
    
    this.label.root.style.padding         = '0px 12px';
    this.label.root.style['padding-left'] = '32px';

    this.checked = false;

    this.uncheck;
    this.check;

    var that = this;
    this.click(function() {
        if(that.checked) {
            that.uncheck();
        } else {
            that.check();
        }
    });
}

Checkbox.prototype = {
    setBackground: function(background) {
        this.nonHighlightBackground = background;
        Input.prototype.setBackground.call(this, background);
    },

    onUncheck: function(callback) {
        var that = this;
        that.uncheck = function() {
            Input.prototype.setBackground.call(that, that.nonHighlightBackground);
            that.circleDiv.setBackground(Colors.buttonUncheckedCircle);
            that.circleDiv.root.style.border = '1px solid ' + Colors.buttonCheckboxCircleBorder;
            callback();
            that.checked = false;
            that.emit('unchecked');
        };
    },

    onCheck: function(callback) {
        var that = this;
        that.check = function() {
            Input.prototype.setBackground.call(that, Colors.buttonCheckedBackground);
            that.circleDiv.setBackground(Colors.buttonCheckedCircle);
            that.circleDiv.root.style.border = 'none';
            callback();
            that.checked = true;
            that.emit('checked');
        };
    },

    removeEvents: function() {
        Button.prototype.removeEvents.call(this);
    },
    __proto__: Button.prototype
};

module.exports = Checkbox;
