// new_ui/lib/colors.js
'use strict';

var lightBlack                  = '#303040',
    lightPink                   = '#f7b8dd',
    lightPurple                 = '#c7b7f7',
    palePurple                  = '#AA6EEB',
    darkPurple                  = '#b197ba',
    darkerPurple                = '#A3779D',
    lightBlue                   = '#b7caf7',
    paleBlue                    = '#7C6CE6',
    blue                        = '#3643A3',
    darkTeal                    = '#446A96',
    darkerTeal                  = '#395585',
    darkererTeal                = '#314973',
    darkerererTeal              = '#2A3D66',
    tealFont                    = '#D7F1F7',
    almostBlue                  = '#3F2191',
    warningRed                  = '#E61953',
    warningRedBorder            = '#D21748',
    lightOrange                 = '#D99B4A',
    darkOrange                  = '#CF743C',
    darkDarkOrange              = '#572211',
    lightGreen                  = '#63E07C',
    darkerLightGreen            = '#27AB5E',

    white                       = '#fafafa';

var sidebarBackground           = darkerTeal,

    sidebarFoldButtonSize       = '30px',
    sidebarFoldButtonFontSize   = '14px',
    sidebarFoldButtonFontColor  = tealFont,
    sidebarFoldButtonBackground = darkTeal,
    sidebarFoldButtonPaddingtop = 5,

    menuButtonHeight            = 30,
    menuButtonIconSize          = '1.2em',
    menuButtonFontColor         = tealFont,
    menuButtonBackground        = darkTeal,
    menuButtonPaddingTop        = 5,

    menuItemFontSize            = '12px',
    menuItemFontColor           = tealFont,
    menuItemBackground          = darkerTeal,
    menuItemActiveBackground    = darkererTeal,

    itemFontSize                = '12px',
    itemFontColor               = tealFont,
    itemBackground              = darkererTeal,

    inputBackground             = darkerererTeal,
    inputFontColor              = tealFont;

var buttonBackground            = inputBackground,
    buttonFontColor             = inputFontColor,
    buttonCheckedBackground     = darkOrange,//lightOrange,
    activeAvatar                = lightOrange;

var warning                     = '#ca4050';//warningRed;

module.exports = {
    lightBlack:                   lightBlack,
    lightPink:                    lightPink,
    lightPurple:                  lightPurple,
    palePurple:                   palePurple,
    darkPurple:                   darkPurple,
    darkerPurple:                 darkerPurple,
    lightBlue:                    lightBlue,
    paleBlue:                     paleBlue,
    blue:                         blue,
    darkTeal:                     darkTeal,
    darkerTeal:                   darkerTeal,
    darkererTeal:                 darkererTeal,
    darkerererTeal:               darkerererTeal,
    tealFont:                     tealFont,
    almostBlue:                   almostBlue,
    lightOrange:                  lightOrange,
    warningRed:                   warningRed,
    warningRedBorder:             warningRedBorder,
    lightGreen:                   lightGreen,
    darkOrange:                   darkOrange,
    darkerLightGreen:             darkerLightGreen,
    darkDarkOrange:               darkDarkOrange,

    white:                        white,

    warning:                      warning,

    sidebarBackground:            '#343749',//sidebarBackground,

    dropdownNew:                  darkerLightGreen,
    dropdownEdit:                 darkOrange,
    dropdownEditInput:            darkDarkOrange,
    dropdownDelete:               warning,

    sidebarFoldButtonSize:        sidebarFoldButtonSize,
    sidebarFoldButtonFontSize:    sidebarFoldButtonFontSize,
    sidebarFoldButtonFontColor:   sidebarFoldButtonFontColor,
    sidebarFoldButtonBackground:  '#343749',//sidebarFoldButtonBackground,
    sidebarFoldButtonPaddingtop:  sidebarFoldButtonPaddingtop,

    menuButtonHeight:             menuButtonHeight,
    menuButtonIconSize:           menuButtonIconSize,
    menuButtonFontColor:          menuButtonFontColor,
    menuButtonBackground:         '#343749',//,menuButtonBackground,
    menuButtonPaddingTop:         menuButtonPaddingTop,

    menuItemFontSize:             menuItemFontSize,
    menuItemFontColor:            menuItemFontColor,
    menuItemBackground:           '#343749',//menuItemBackground,
    menuItemActiveBackground:     '#207090',//menuItemActiveBackground,

    itemFontSize:                 itemFontSize,
    itemFontColor:                '#202020',//itemFontColor,
    itemBackground:               '#eaeaea',//itemBackground,

    inputBackground:              '#fff',//inputBackground,
    inputFontColor:               '#202020',//inputFontColor,

    buttonBackground:             '#fff',//inputBackground,
    buttonFontColor:              '#202020',//inputFontColor,
    buttonCheckedBackground:      '#fff',//buttonCheckedBackground,
    buttonUncheckedCircle:        '#fff',
    buttonCheckedCircle:          '#3090A0',
    buttonCheckboxCircleBorder:   '#207090',
    activeAvatar:                 '#207090'//activeAvatar
};
