// new_ui/lib/element.js
'use strict';

var Emitter = require('emitter');

function Element(type) {
    Emitter.call(this);
    this.root = document.createElement(type || 'div');
}

Element.prototype = {
    appendTo: function(container) {
        if(container instanceof Element) {
            return container.appendChild(this);
        }

        container.appendChild(this.root);
    },

    setLabel: function(label) {
        this.root.innerHTML = label;
    },

    getLabel: function() {
        return this.root.innerHTML;
    },

    appendChild: function(container) {
        if(!this.children) {
            this.children = [];
        }

        if(container instanceof Element) {
            var index = this.children.indexOf(container);
            if(index !== -1) {
                this.children.splice(index, 1);
            }

            this.children.push(container);
        }

        if(container.root) {
            return this.root.appendChild(container.root);
        }

        this.root.appendChild(container);
    },

    insertChildAt: function(container, index) {
        if(!this.children || this.children.length === 0) {
            return this.appendChild(container);
        }

        if(index > this.children.length) {
            return this.appendChild(container);
        } else if(index < 0) {
            index = 0;
        }

        if(container.root) {
            var existing = this.children.indexOf();
            if(existing !== 1) {
                this.children.splice(existing, 1);
            }
            
            this.children.splice(index, 0, container);
            return this.root.insertBefore(container.root, this.root.children[index]);
        }

        this.root.insertBefore(container, this.root.children[index]);
    },

    removeChild: function(child) {
        if(!this.children) {
            try {
                if(child.root) {
                    this.root.removeChild(child.root);
                } else {
                    this.root.removeChild(child)
                }
            } catch(err) {
                console.error(err);
            }

            return;
        }

        var index = this.children.indexOf(child);
        if(index === -1) {
            try {
                this.root.removeChild(child);
            } catch(err) {
                console.error(err);
            }

            return;
        }

        this.children.splice(index, 1);
        this.root.removeChild(child.root);
    },

    createFoldButton: function() {
        if(!this.buttons) {
            this.buttons = [];
        }

        var b = new FoldButton();
        this.buttons.push(b);

        return b;
    },

    createButton: function() {
        if(!this.buttons) {
            this.buttons = [];
        }

        var b = new Button();
        this.buttons.push(b);

        return b;
    },

    deleteAllButtons: function() {
        if(!this.buttons) {
            return;
        }

        this.buttons.forEach(function(button) {
            button.removeEventListener();
        });
    },

    setWidth: function(width) {
        if(typeof width === 'string') {
            this.root.style.width = width;
        } else if(typeof width === 'number') {
            this.root.style.width = width + 'px';
        }

        this.width = parseInt(width);
    },

    getWidth: function() {
        return this.root.offsetWidth || this.width || 0;
    },

    setHeight: function(height) {
        if(typeof height === 'string') {
            this.root.style.height = height;
        } else if(typeof height === 'number') {
            this.root.style.height = height + 'px';
            this.height = parseInt(height);
        }
    },

    getHeight: function() {
        return this.root.offsetHeight || this.height || 0;
    },

    setBackground: function(background) {
        this.root.style['background-color'] = background;
        this.background = background;
    },

    getBackground: function() {
        return this.background;
    },

    setLeft: function(left) {
        if(typeof left === 'string') {
            this.root.style.left = left;
        } else if(typeof left === 'number') {
            this.root.style.left = left + 'px';
        }

        this.left = parseInt(left);
    },

    getLeft: function() {
        return this.left || 0;
    },

    setTop: function(top) {
        if(typeof top === 'string') {
            this.root.style.top = top;
        } else if(typeof top === 'number') {
            this.root.style.top = top + 'px';
        }

        this.top = parseInt(top);
    },

    getTop: function() {
        return this.top || 0;
    },

    hide: function() {
        if(this.root.style.display !== 'none') {
            this.currentStyle = this.root.style.display;
        }

        this.root.style.display = 'none';
    },

    show: function() {
        this.root.style.display = this.currentStyle || 'block';
    },

    destroy: function() {
        if(this.removeEvents) {
            this.removeEvents();
        }

        if(this.children) {
            this.children.forEach(function(child) {
                child.destroy();
            });
        }

        if(this.root.parentElement) {
            this.root.parentElement.removeChild(this.root);
        }
    },

    __proto__: Emitter.prototype
};

module.exports = Element;
