// new_ui/lib/foldable.js
'use strict';

var Element = require('./element.js'),
    Tween   = require('./tween.js');

var easeOutCirc = Tween.easeOutCirc;

function Foldable(width, positional) {
    if(!width) {
        throw new Error('Creating foldable without width.');
    }

    if(typeof width !== 'number') {
        throw new Error('Width given must be a number.');
    }

    Element.call(this);
    this.root = document.createElement('div');

    if(positional) {
        this.setWidth(width);
    } else {
        this.setWidth(0);
    }

    this.folded        = true;
    this.foldableWidth = width;
    this.children      = [];
}

Foldable.prototype = {
    fold: function(onDone) {
        var that = this;
        return new Promise(function(fulfill, reject) {
            that.children.forEach(function(child) {
                if(child instanceof Foldable) {
                    child.fold();
                }
            });

            if(that.currentTween) {
                that.currentTween.stop();
            }

            var currentWidth = that.getWidth();
            var destination  = currentWidth;

            that.currentTween = easeOutCirc(destination, 250, function(width) {
                that.setWidth(currentWidth - width);
            }, function(){if(onDone){onDone();} fulfill();});

            that.folded = true;
            that.emit('folded');
        });
    },

    unfold: function(onDone) {
        var that = this;
        return new Promise(function(fulfill, reject) {
            that.children.forEach(function(child) {
                if(child instanceof Foldable) {
                    if(child.wasUnfolded) {
                        child.unfold();
                    }
                }
            });

            if(that.currentTween) {
                that.currentTween.stop();
            }

            var currentWidth = that.getWidth();
            var destination  = that.foldableWidth - currentWidth;

            that.currentTween = easeOutCirc(destination, 250, function(width) {
                that.setWidth(currentWidth + width);
            }, function(){if(onDone){onDone();} fulfill();});

            that.folded = false;
            that.emit('unfolded');
        });
    },

    positionalFold: function() {
        this.children.forEach(function(child) {
            if(child instanceof Foldable) {
                child.fold();
            }
        });

        if(this.currentTween) {
            this.currentTween.stop();
        }

        var currentLeft = this.getLeft();
        var destination = currentLeft;

        if(destination === 0) {
            this.folded = true;
            return;
        }

        var that = this;
        this.currentTween = easeOutCirc(destination, 250, function(left) {
            that.setLeft(currentLeft - left);
        });

        this.folded = true;
    },

    positionalUnfold: function() {
        this.children.forEach(function(child) {
            if(child instanceof Foldable) {
                child.unfold();
            }
        });

        if(this.currentTween) {
            this.currentTween.stop();
        }

        var currentLeft = this.getLeft();
        var destination = this.foldableWidth - currentLeft;

        var that = this;
        this.currentTween = easeOutCirc(destination, 250, function(left) {
            that.setLeft(currentLeft + left);
        });

        this.folded = false;
    },

    positionalInvert: function() {
        if(this.folded) {
            this.positionalUnfold();
        } else {
            this.positionalFold();
        }
    },

    invert: function(onDone) {
        if(this.folded) {
            return this.unfold(onDone);
        }

        return this.fold(onDone);
    },

    __proto__: Element.prototype
};

module.exports = Foldable;
