// new_ui/lib/input.js
'use strict';

var Element = require('./element.js'),
    Colors  = require('./colors.js');

function Input() {
    this.changes = [];
    this.inputs  = [];

    Element.call(this);

    this.input    = new Element('input');
    this.label    = new Element('div');

    this.inputDiv = new Element('div');
    //this.inputDiv.appendChild(this.input);

    this.input.root.style.display = 'inline-block';
    this.label.root.style.display = 'inline-block';
    this.label.setWidth('45%');
    this.input.setWidth('55%');
    this.input.setBackground(Colors.inputBackground);
    this.input.root.style.border           = 'none';
    this.input.root.style.color            = Colors.inputFontColor;
    this.label.root.style['text-align']    = 'left';
    this.input.root.style['text-align']    = 'left';
    this.input.root.style.padding          = '8px';

    var that = this;
    this.onChange(function() {
        that.input.root.blur();
    });

    this.appendChild(this.label);
    this.appendChild(this.input);
}

Input.prototype = {
    refresh: function() {
        if(!this.updateValue) {
            return;
        }
        
        this.updateValue();
    },
    
    defaultValue: function(callback) {
        if(!callback && typeof callback !== 'function') {
            throw new Error('Trying to set callback which is not a function');
        }
        
        var that = this;
        this.updateValue = function() {
            that.setValue(callback());
        };

        this.updateValue();
    },

    setLabel: function(label) {
        this.label.root.style['margin-bottom'] = '8px';
        this.label.root.innerHTML = label;
    },

    setValue: function(value) {
        this.input.root.value = value;
    },

    getValue: function() {
        return this.input.root.value;
    },

    onChange: function(callback) {
        this.changes.push(callback);
        this.input.root.addEventListener('change', callback);
    },

    onInput: function(callback) {
        this.inputs.push(callback);
        this.input.root.addEventListener('input', callback);
    },

    replaceChange: function(callback, newCallback) {
        var index = changes.indexOf(callback);
        this.input.root.removeEventListener('change', changes[index]);
        this.input.root.addEventListener('change', newCallback);

        changes[index] = newCallback;
    },

    replaceInput: function(callback, newCallback) {
        var index = inputs.indexOf(callback);
        this.input.root.removeEventListener('input', inputs[index]);
        this.input.root.addEventListener('input', newCallback);

        inputs[index] = newCallback;
    },

    removeEvents: function() {
        this.inputs.forEach(function(callback) {
            this.input.root.removeEventListener('input', callback);
        }, this);

        this.changes.forEach(function(callback) {
            this.input.root.removeEventListener('change', callback);
        }, this);
    },

    __proto__: Element.prototype
}

module.exports = Input;
